lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -3707,3 +3733,75 @@ def get_test_data(delta=0.05):
     Y = Y * 10
     Z = Z * 500
     return X, Y, Z
+
+
+class _Quaternion:
+    """
+    Quaternions
+    consisting of scalar, along 1, and vector, with components along i, j, k
+    """
+
+    def __init__(self, scalar, vector):
+        self.scalar = scalar
+        self.vector = np.array(vector)
+
+    def __neg__(self):
+        return self.__class__(-self.scalar, -self.vector)
+
+    def __mul__(self, other):
+        """
+        Product of two quaternions
+        i*i = j*j = k*k = i*j*k = -1
+        """
+        return self.__class__(
+            self.scalar*other.scalar - np.dot(self.vector, other.vector),
+            self.scalar*other.vector + self.vector*other.scalar
+            + np.cross(self.vector, other.vector))
+
+    def __eq__(self, other):
+        return (self.scalar == other.scalar) and (self.vector == other.vector).all
+
+    def __repr__(self):
+        return "_Quaternion({}, {})".format(repr(self.scalar), repr(self.vector))
+
+    @classmethod
+    def from_to(cls, r1, r2):
+        """The quaternion that rotates vector r1 to vector r2."""
+        k = np.cross(r1, r2)
+        nk = np.linalg.norm(k)
+        th = np.arctan2(nk, np.dot(r1, r2))
+        th = th/2
+        if nk == 0:  # r1 and r2 are parallel or anti-parallel
+            if np.dot(r1, r2) < 0:
+                q = cls(0, [1, 0, 0])  # axis of the circle that r1 and r2 are on
```

Comment:
This rotates 180 deg about the i axis, where you actually want to rotate about a vector that is orthogonal to r1 and r2. I was just fixing this over in scipy's implementation, go take a look at the logic there:
https://github.com/scipy/scipy/blob/main/scipy/spatial/transform/_rotation.pyx#L3483-L3511

Reply:
Please also make sure you have tests for these two special cases.

Reply:
While that may be the case for its current use, I think we may end up using this class more broadly to perform 3D operations in the library. The corrected logic is lightweight, let's get it right now so it doesn't bite us later - perfect 180 degree rotations end up being a pretty large fraction of use cases in my experience.

Reply:
> [...] perfect 180 degree rotations end up being a pretty large fraction of use cases in my experience.

So what would you want to return as a result? For the case of opposing vectors, there is a whole great circle of normal vectors that could work as rotation axis, taking r1 to r2. All rotate over 180 degrees, which one is shortest?

Reply:
Well, all of them are! I think it’s fine to pick an arbitrary one for this edge case.

Reply:
`    q = cls(0, np.linalg.svd([r1, r2]). Vh[-1])`
A suitable sledgehammer to crack that nut.

Reply:
Longer, but perhaps more suitable:
```
warnings.warn("Rotation defined by anti-parallel vectors is ambiguous")
k = np.zeros(3)
k[np.argmin(r1*r1)] = 1  # basis vector most perpendicular to r1-r2
k = np.cross(r1, k)
k = k / np.linalg.norm(k)  # unit vector normal to r1-r2
q = cls(0, k)
```